Fix more leaks to default copy of bundler

deivid-rodriguez opened the pull request:
### What was the end-user problem that led to this PR?

The problem was that in some places, it was still possible to end up requiring files in a different copy of bundler (the default copy). I noticed this when I removed a rubygems monkeypatch from the test suite that was preventing the default copy of bundler from being activated when requiring files.

This thing:

https://github.com/bundler/bundler/blob/e1c518363641208429f397170354054b3d28effd/spec/support/hax.rb#L15-L20

### What was your diagnosis of the problem?

My diagnosis was that I should use relative requires wherever they were missing.

### What is your fix for the problem, implemented in this PR?

My fix is to remove the rubygems hack, migrate the rest of the internal requires to be relative, and also introduce some hacks on our specs to make sure we never load the incorrect copy of bundler.

I think this PR should fix the issues in https://github.com/rubygems/rubygems/pull/2863.

deivid-rodriguez commented:
Unfortunately this patch seems to introduce some flakyness on a single realworld spec 

https://github.com/bundler/bundler/blob/e1c518363641208429f397170354054b3d28effd/spec/realworld/double_check_spec.rb#L1-L40

And I'm not yet sure why :grimacing:.

deivid-rodriguez commented:
Actually, nevermind, the flakyness is already present in master, so it's unrelated to this PR.

I'll investigate how the flaky got started separately.

deivid-rodriguez commented:
I still want to review some things in this PR, so please don't merge it just yet :)

hsbt commented on spec/support/path.rb:
```diff
@@ -108,10 +108,6 @@ def lib_path(*args)
       tmp("libs", *args)
     end
 
-    def bundler_path
```
I think we restore `bundler_path` and use it instead like `require "../../lib/bundler"`. Because of the directory structure of the ruby core repository for spec files are different from this repository.

deivid-rodriguez replied:
The reason I removed it is because it was duplicated with the `lib` helper method. But I agree with your suggestion, so I will switch to it :+1:. Thanks!

deivid-rodriguez replied:
I moved the `lib` helper to `spec/support/path.rb`, since it had the exact same implementation of this method. Now the `lib` helper is used everywhere.

deivid-rodriguez commented on spec/install/gemfile/groups_spec.rb:
```diff
@@ -333,7 +333,7 @@
 
       ruby <<-R
-        require "bundler"
+        require "#{__dir__}/../../../lib/bundler"
```
`require_relative` wasn't working here, I guess because the command is run from a different folder form the current test folder.

hsbt replied:
I see. Thanks.

deivid-rodriguez replied:
Now everything is more consistent because the path helpers (independent from the test folder structure) are used everywhere. :+1:

deivid-rodriguez commented:
@hsbt I addressed your feedback and now hopefully all paths are independent from the specific folder structure where the tests are run, thanks for pointing that out :+1:.

I still want to improve a few other things in this PR, so I'll let you know when it's ready.

deivid-rodriguez commented:
I'm now happy with this PR, @hsbt. You can have another look :)

deivid-rodriguez commented:
@hsbt If you're ok, I'd like to merge this.

hsbt commented:
@deivid-rodriguez I'm going to try to merge this pull-request into ruby repo soon.

deivid-rodriguez commented:
Thank you @hsbt, let me know how it goes! Hopefully it allows removing some of those `:ruby_core` tags :)

hsbt commented:
We need to rebase this after merging https://github.com/bundler/bundler/pull/7303/

deivid-rodriguez commented:
> We need to rebase this after merging #7303

Yes, I'll take care of that :+1:.

deivid-rodriguez commented:
@hsbt This one is ready, can I merge it?

deivid-rodriguez commented:
@bundlerbot r=hsbt
